server/app/app.py

```diff
@@ -30,13 +193,25 @@ def create_app(self):
         # Config
         SECRET_KEY = os.environ.get("CXG_SECRET_KEY", default="SparkleAndShine")
         self.app.config.update(SECRET_KEY=SECRET_KEY)
-        self.app.config.update(SCRIPTS=[])
 
-        resources = get_api_resources()
-        self.app.register_blueprint(webapp.bp)
-        self.app.register_blueprint(resources.blueprint)
-        self.app.add_url_rule("/", endpoint="index")
+        self.app.register_blueprint(webbp)
+
+        api_version = "/api/v0.2"
+        if data:
+            bp_api = Blueprint("api", __name__, url_prefix=api_version)
+            resources = get_api_resources(bp_api)
+            self.app.register_blueprint(resources.blueprint)
+            self.app.add_url_rule("/", "index", index)
+
+        if app_config.dataroot:
+            bp_api = Blueprint("api_dataset", __name__, url_prefix="/<dataset>" + api_version)
+            resources = get_api_resources(bp_api)
+            self.app.register_blueprint(resources.blueprint)
+            self.app.add_url_rule("/<dataset>/", 'dataset_index', dataset_index)
+            self.app.add_url_rule("/<dataset>/static/<path:therest>", "static_redirect", static_redirect)
+            self.app.add_url_rule("/<dataset>/favicon.png", "favicon_redirect", favicon_redirect)
 
-    def attach_data(self, data, title="Demo", about=""):
-        self.app.config.update(DATASET_TITLE=title, ABOUT_DATASET=about)
         self.app.data = data
+        self.app.annotations = annotations
+        self.app.app_config = app_config
+        self.app.matrix_data_cache_manager = MatrixDataCacheManager()
```

Comment:
why unconditionally allocate a cache manager when `rest_get_data_adaptor` will only use it in the `--dataroot` mode?   

I would either always use the cache (ie, `current_app.data` becomes the name or location of the default data, rather than the pre-loaded adaptor), *OR* don't bother to create the cache manager when in single-dataset mode.

I kind of like the idea of always using the cache manager to load, so we have exactly one code path for loading all data.

This would probably require creating the CacheManager in launch.py, and passing it down to Server.__init__ (because single dataset mode wants to do some early checks such as file size).   But I don't feel strongly - just need to more cleanly separate these modes here and in `rest_get_data_adaptor`

Reply:
The matrix_data_cache_manager is now passed into the Server, and preloaded if a default dataset was provided.